File: Sol17_Letter_Combination_of_a_Phone_Number/Solution17.py
class Solution:
    def letterCombinations(self, digits):
        if len(digits) == 0:
            return []

        mp = [
            "", "", "abc", "def", "ghi",
            "jkl", "mno", "pqrs", "tuv", "wxyz"
        ]

        ans = [""]

        for i in range(len(digits)):
            let = mp[int(digits[i])]
            temp = []

            for s in ans:
                for j in range(len(let)):
                    temp.append(s + let[j])

            ans = temp

        return ans


# Example / Testing
if __name__ == "__main__":

    solution = Solution()

    # Example 1
    digits = "23"
    result = solution.letterCombinations(digits)

    print("Input:", digits)
    print("Output:", result)

    # Example 2
    digits = "2"
    result = solution.letterCombinations(digits)

    print("\nInput:", digits)
    print("Output:", result)

    # Example 3
    digits = ""
    result = solution.letterCombinations(digits)

    print("\nInput: empty")
    print("Output:", result)

File: Sol17_Letter_Combination_of_a_Phone_Number/Solution17.java
import java.util.*;

class Solution {

    public List<String> letterCombinations(String digits) {

        // If input is empty, return empty list
        if (digits.length() == 0) {
            return new ArrayList<>();
        }

        // Phone keypad mapping
        String[] map = {
            "", "", "abc", "def", "ghi",
            "jkl", "mno", "pqrs", "tuv", "wxyz"
        };

        List<String> ans = new ArrayList<>();

        // Start with an empty combination
        ans.add("");

        // Go through each digit
        for (int i = 0; i < digits.length(); i++) {

            // Get letters for current digit
            String let = map[digits.charAt(i) - '0'];

            List<String> temp = new ArrayList<>();

            // Take every existing combination
            for (String s : ans) {

                // Add every possible letter
                for (int j = 0; j < let.length(); j++) {
                    temp.add(s + let.charAt(j));
                }
            }

            // Move the new combinations into ans
            ans = temp;
        }

        return ans;
    }


    // =========================
    // EXAMPLE / TESTING
    // =========================

    public static void main(String[] args) {

        Solution obj = new Solution();

        // Example 1
        String digits1 = "23";

        List<String> result1 = obj.letterCombinations(digits1);

        System.out.println("Input: " + digits1);
        System.out.println("Output: " + result1);


        // Example 2
        String digits2 = "2";

        List<String> result2 = obj.letterCombinations(digits2);

        System.out.println("\nInput: " + digits2);
        System.out.println("Output: " + result2);


        // Example 3
        String digits3 = "";

        List<String> result3 = obj.letterCombinations(digits3);

        System.out.println("\nInput: empty");
        System.out.println("Output: " + result3);
    }
}